libsignal-protocol/src/keys/pre_key.rs

```diff
@@ -0,0 +1,54 @@
+use crate::{
+    errors::FromInternalErrorCode, keys::KeyPair, raw_ptr::Raw, Buffer,
+};
+use failure::Error;
+use std::{io::Write, ptr};
+
+pub struct PreKey {
+    pub(crate) raw: Raw<sys::session_pre_key>,
+}
+
+impl PreKey {
+    pub fn new(id: u32, key_pair: &KeyPair) -> Result<PreKey, Error> {
+        unsafe {
+            let mut raw = ptr::null_mut();
+            sys::session_pre_key_create(&mut raw, id, key_pair.raw.as_ptr())
+                .into_result()?;
+
+            Ok(PreKey {
+                raw: Raw::from_ptr(raw),
+            })
+        }
+    }
+
+    pub fn serialize<W: Write>(&self, mut writer: W) -> Result<(), Error> {
```

Comment:
So, it is just a question, why not using
`mut writer: impl Write` ?

Is there any difference?

Reply:
I think it's mainly a stylistic thing. I like using `impl Trait` in return position, but haven't gotten used to seeing/putting it in the argument position.

For all intents and purposes, the two are identical.